marketplaceItems/textToSpeech/hifiScript/appData/ui/js/textToSpeech_ui.js

```diff
@@ -0,0 +1,524 @@
+
+    emitAppSpecificEvent("translateTextThenSpeak", {
+        textToSpeak: inputText.value
+    });
+}
+
+
+// Sends a message to the App JS to tell it to stop any active TTS speech.
+function stopSpeech() {
+    emitAppSpecificEvent("stopSpeech");
+}
+
+
+// Translates a Google TTS API Voice name (i.e. "en-US-Wavenet-A") to something more readable
+function getReadableVoiceName(voiceName) {
+    var suffix = voiceName.split("-");
+    suffix = suffix[(suffix.length - 1)];
+
+    var readableVoiceName = `Voice ${suffix}`;
+
+    return readableVoiceName;
+}
+
+
+// Called when a user clicks on a button to select a voice gender (after selecting a language):
+// 1. Removes the "selectedButton" button class from all gender choice buttons.
+// 2. Adds the "selectedButton" button class to the selected gender.
+// 3. Clears all buttons from the `voiceButtonsContainer` (which holds all voice choices for this combination
+//     of language and gender).
+// 4. Determines the user's selected language in the UI.
+// 5. Sorts the pre-filled `voicesObject` by all possible voices (for this combination of language and gender).
+// 6. For each voice in (5), creates a clickable button for that voice choice.
+// 7 (optional). If the caller sets `isStartingUp = true`, then this code won't call `changeVoiceButtonClicked()`.
+//     Otherwise, that'll happen, and the user will be forced to choose the default voice associated with that gender.
+function genderButtonClicked(gender, isStartingUp) {
+    var genderButtons = document.getElementsByClassName("genderButton");
+    for (var i = 0; i < genderButtons.length; i++) {
+        genderButtons[i].classList.remove("selectedButton");
+
+        if (genderButtons[i].getAttribute("data-gender") === gender) {
+            genderButtons[i].classList.add("selectedButton");
+        }
+    }
+
+    var voiceButtonsContainer = document.getElementById("voiceButtonsContainer");
+    voiceButtonsContainer.innerHTML = "";
+
+    var selectedLanguage;
+    var languageButtons = document.getElementsByClassName("languageButton");
+    for (var i = 0; i < languageButtons.length; i++) {
+        if (languageButtons[i].classList.contains("selectedButton")) {
+            selectedLanguage = languageButtons[i].getAttribute("data-language");
+            break;
+        }
+    }
+
+    if (!selectedLanguage) {
+        console.error("Could not find selectedButton language in the DOM!");
+        return;
+    }
+
+    var possibleVoices = voicesObject[selectedLanguage][gender].sort();
+    for (var i = 0; i < possibleVoices.length; i++) {
+        var div = document.createElement("div");
+        div.classList.add("changeVoiceButtonContainer");
+        div.addEventListener("click", function (event) {
+            previewVoiceButtonClicked(event.target.childNodes[0].getAttribute("data-voiceName"), event.target.childNodes[0].getAttribute("data-targetLanguageCode"), event.target.childNodes[0].getAttribute("data-voiceGender"));

```

Comment:
maybe one variable to event.target.childNodes[0] since it is called 3 times in a row

Reply:
👍